qiime2/sdk/usage.py

```diff
@@ -0,0 +1,349 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2020, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+import abc
+import os
+import re
+import types
+import tempfile
+import uuid
+import zipfile
+
+from qiime2 import sdk
+
+# TODO: docstrings
+
+
+class UsageAction:
+    def __init__(self, plugin_id: str, action_name: str):
+        if plugin_id == '':
+            raise ValueError('Must specify a value for plugin_id.')
+
+        if action_name == '':
+            raise ValueError('Must specify a value for action_name.')
+
+        self.plugin_id = plugin_id
+        self.action_name = action_name
+        self._plugin_manager = sdk.PluginManager()
+
+    def __repr__(self):
+        return 'UsageAction(plugin_id=%r, action_name=%r)' %\
+            (self.plugin_id, self.action_name)
+
+    def get_action(self):
+        plugin = self._plugin_manager.get_plugin(id=self.plugin_id)
+        # TODO: should this validation be pushed up into
+        # plugin.py or action.py?
+        try:
+            action_f = plugin.actions[self.action_name]
+        except KeyError:
+            raise KeyError('No action currently registered with '
+                           'name: "%s".' % (self.action_name,))
+        return (action_f, action_f.signature)
+
+    def validate(self, inputs, outputs):
+        if not isinstance(inputs, UsageInputs):
+            raise TypeError('Must provide an instance of UsageInputs.')
+        if not isinstance(outputs, UsageOutputNames):
+            raise TypeError('Must provide an instance of UsageOutputNames.')
+
+        _, sig = self.get_action()
+
+        inputs.validate(sig)
+        outputs.validate(sig)
+
+
+class UsageInputs:
+    def __init__(self, **kwargs):
+        self.values = kwargs
+
+    def __repr__(self):
+        return 'UsageInputs(**%r)' % (self.values,)
+
+    def validate(self, signature):
+        provided = set(self.values.keys())
+        inputs, params = signature.inputs, signature.parameters
+        exp_inputs = {k for k, v in inputs.items() if not v.has_default()}
+        exp_params = {k for k, v in params.items() if not v.has_default()}
+
+        missing = exp_inputs - provided
+        if len(missing) > 0:
+            raise ValueError('Missing input(s): %r' % (missing, ))
+
+        missing = exp_params - provided
+        if len(missing) > 0:
+            raise ValueError('Missing parameter(s): %r' % (missing, ))
+
+        extra = provided - exp_inputs - exp_params
+        if len(extra) > 0:
+            raise ValueError('Extra input(s) or parameter(s): %r' %
+                             (extra, ))
+
+    def prepare(self, signature, scope):
```

Comment:
would `bind` be a better word for this?

Reply:
How about `build_opts`?

Reply:
Right now there is no real typechecking in the base `Usage` class - `ExecutionUsage` performs typechecking in a trial-by-fire approach, which is better than nothing, but as it stands, unless a usage example is invoked with `ExecutionUsage`, it is "unproven." I think that `build_opts` might be a good place to beef up typechecking, let's chat about this IRL.

Reply:
Maybe the trial-by-fire isn't so bad. I need to stew on this, since the trade-off is the factory business becomes a little superfluous with the purposes of typechecking (its probably a lot easier to check types with materialized data, rather than factories).